enhancements/sig-policy/85-gatekeeper-policy-integration/README.md

```diff
@@ -0,0 +1,257 @@
+`ConstraintTemplate`/constraints need to be removed when the OCM `Policy` object is deleted and if applicable, when the
+`Policy` object is updated. To be able to track which Gatekeeper constraints belong to a `Policy` object, a new label
+should be added to the Gatekeeper `ConstraintTemplate` objects and constraints of
+`policy.open-cluster-management.io/policy` with the policy name. Note that a finalizer will also need to be added to the
+`governance-policy-framework` `Deployment` like the Configuration Policy controller does to prevent the Template Sync
+from being deleted before the finalizer can be handled.
+
+The Template Sync controller would also need to generate a status event when it creates or updates Gatekeeper
+`ConstraintTemplate` objects since they are implementation details for the Gatekeeper constraints and don't provide any
+validation on their own.
+
+If the Gatekeeper `ConstraintTemplate` includes invalid rego syntax, a Gatekeeper validation webhook automatically
+rejects the requst as shown below. When this occurs, the OCM `Policy` will be set to non-compliant and include the error
+returned by the validation webhook.
+
+```bash
+$ kubectl apply -f bad-constraint-template.yaml
+Error from server (1 error occurred: templates["admission.k8s.gatekeeper.sh"]["K8sRequiredLabels"]:7: rego_parse_error: unexpected eof token
+          mis
+            ^): error when creating "temp6.yaml": admission webhook "validation.gatekeeper.sh" denied the request: 1 error occurred: templates["admission.k8s.gatekeeper.sh"]["K8sRequiredLabels"]:7: rego_parse_error: unexpected eof token
+          mis
+            ^
+```
+
+#### New Controller
+
+In the `governance-policy-framework-addon` controller manager, a new controller called `gatekeeper-constraint` would be
+added. This is a bit tricky because each Gatekeeper `ConstraintTemplate` produces a new custom resource definition (CRD)
+and thus the watches aren't known at controller instantiation time.
+
+To handle this, the controller would watch all `Policy` objects in the managed cluster namespace. During the reconcile,
+the controller will detect if Gatekeeper constraints are managed by the policy. If so, the
+[dynamic watch](https://github.com/stolostron/kubernetes-dependency-watches) library will be used to watch the
+individual Gatekeeper constraints and trigger reconciles on the `Policy` parent object whenever they have status
+updates.
+
+The new controller would also examine the `Policy` status before sending a status event to avoid sending duplicate
+status events. It's possible that duplicate status events could be sent if the Status Sync controller is overwhelmed and
+the Gatekeeper constraint has multiple reconciles in succession. This is considered acceptable because the Status Sync
+controller filters out duplicate status events.
+
+### Risks and Mitigation
+
+### Open Questions [optional]
```

Comment:
ConstraintTemplates and Constraints are `v1beta1` in Gatekeeper, I wonder if we need to carefully consider how to handle if/when they change version?

Reply:
Good point. `ConstraintTemplate` seems to be already `v1` in newer Gatekeeper versions. I'll add a note to support both `v1` and `v1beta1`.

Constraints on the other hand, I think we will just need to cross that bridge when it does get to v1 since it's still v1beta1:
https://github.com/open-policy-agent/frameworks/blob/5e694bda2ffa6f35fa6b82af26d3c47b043b7040/constraint/pkg/client/crds/crds.go#L46